### ISSUE TITLE
Support PostgreSQL's JSON data type(s)

### ISSUE THREAD

**Issue by amotl** [crate/crate#11347]
<details>
  <summary>Original issue description:</summary>

Hi there,

in order to insert JSON data into the database, CrateDB's documentation says:

> You can insert objects using JSON strings. To do this, you must type cast the string to an object with an implicit cast (i.e., passing a string into an object column) or an explicit cast (i.e., using the `::object` syntax). [1]

Relating to [2], I wonder if CrateDB could provide a `::json` alias to its type casting implementation underpinning the `::object` syntax notation in order to implement at least _some_ features of the official PostgreSQL JSON support [3,4], maybe even without further ado? In the case of [2], it might improve the compatibility with how Confluent's JDBC Sink Connector for Kafka works [5].

@asavran reports at [2]:

> We tried to execute the same `INSERT` command manually by substituting the `::json` with `::object` and it worked. Is there a workaround for this?

I know that things might turn out to be more complicated more often than not, so I am just creating this as a basis for discussions.

With kind regards,
Andreas.

</details>

Updated description:

---

**Problem description**

Kafka-connect-JDBC adds explicit `::json` casts for target columns of type `object` (exposed in `pg_catalog` as `json`).

This currently breaks because `json` isn't available as type.

The reason for kafka-connect-JDBC doing that is that otherwise PostgreSQL errors when using prepared statements. (http://sqlfiddle.com/#!17/7e29e/12, https://github.com/pgjdbc/pgjdbc/issues/265, https://github.com/confluentinc/kafka-connect-jdbc/pull/875)

**Proposed potential solutions so far**

1. Add a `json` alias for `object`
2. Add a dedicated new `json` type (without storage and cast support from `text → json`, `json -> object`, `object → json`)
3. Make kafka-connect-jdbc CrateDB aware or configurable, so that no explicit casts are added for `json` columns (see https://github.com/crate/crate/issues/11347#issuecomment-839751709) 

Estimate is for option 2, which is the first choice here.

[1]: https://crate.io/docs/crate/reference/en/4.5/general/ddl/data-types.html#json
[2]: https://community.crate.io/t/insert-a-column-as-json-using-kafka-jdbc-sink/663
[3]: https://www.postgresql.org/docs/current/datatype-json.html
[4]: https://www.postgresql.org/docs/current/functions-json.html
[5]: https://docs.confluent.io/kafka-connect-jdbc/current/sink-connector/index.html#jdbc-sink-data-mapping

**Comment by dobe** [comment]
thumbs up for ::json alias for ::object, this would  make the kafka jdbc sink work

**Comment by proddata** [comment]
PG has a different syntax to interact with `json` fields, though we already return a `pg json (oid 114)` on  `SELECT` on an `OBJECT` anyway through jdbc, so if that solves it for Kafka Connect / Confluent, I'd say this should be prioritised.

**Comment by dobe** [comment]
just to clarify: as shown in https://community.crate.io/t/insert-a-column-as-json-using-kafka-jdbc-sink/663 the jdbc sink generates a cast in the statement. the datatype sent over the wire is a string in this case. 

in the long run we can also support the whole json(b) operators on the crate object type to have better compatibility with pg.

**Comment by mfussenegger** [comment]
Could we add some more information to this, so that we've a clear picture about the problem:

- Which kafka-connect-jdbc version is used
- Is it used with the PostgreSQL JDBC client or with the CrateDB JDBC client? Which version?
- How's the connect-jdbc connector configured? (schema registry, insert mode, pk mode, value serializer, etc.)


Adding a `json` alias may be a suitable solution, but extending the type system tends to have far reaching implications so we better be sure it's the right way to go.

**Comment by mfussenegger** [comment]
> are there any drawbacks doing so?


Not sure.

One thing is that our `object` is conceptionally closer to `jsonb` than `json`.
`json` in PostgreSQL has some semantics that our `object` type doesn't have, like preserving order of keys, preserving duplicate keys and preserving whitespaces. 

It is one thing to use `json` for serialization to clients, another to say that `json == object`.

Anyways - I'm not saying that we shouldn't do it. Just that we have to think about the implications. Cast semantics come to mind.

It also sounds like the current use case is really "I want to be able to ingest into object columns using the kafka-connect-jdbc sink" - this story could have multiple solutions and is different from "Add a JSON alias". I'd prefer if we stick to outlining the problem first before we get hung up on one potential solution of many.

**Comment by hammerhead** [comment]
The issue can be reproduced by following our [Kafka Connect Guide](https://crate.io/docs/crate/howtos/en/latest/integrations/kafka-connect.html) and adjusting the there mentioned `cratedb_connector.json` as follows:
```json
{
  "name": "cratedb-connector",
  "config": {
    "connector.class": "io.confluent.connect.jdbc.JdbcSinkConnector",
    "topics": "metrics",
    "connection.url": "jdbc:postgresql://localhost:5432/doc?user=crate",
    "tasks.max": 3,
    "insert.mode": "insert",
    "table.name.format": "metrics",
    "transforms": "HoistPayload",
    "transforms.HoistPayload.field": "payload",
    "value.converter": "org.apache.kafka.connect.storage.StringConverter",
    "transforms.HoistPayload.type": "org.apache.kafka.connect.transforms.HoistField$Value"
  }
}
```

Confluent Connect derives the necessity to perform an explicit cast based on the column definition, see their [PostgreSqlDatabaseDialect.java](https://github.com/confluentinc/kafka-connect-jdbc/blob/47b7177d4fe4651450b54f7f9c176f2db92a8ca6/src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java#L419).

> PG has a different syntax to interact with json fields, though we already return a pg json (oid 114) on SELECT on an OBJECT anyway through jdbc

Returning the column as `json` in `pg_catalog` is exactly what causes Confluent Connect to add the `::json` cast. In [this commit](https://github.com/confluentinc/kafka-connect-jdbc/commit/74fad2d5683c0f5f77784dc4f3ac94fd10b6fa59) the cast was added:
> the `PostgreSqlDatabaseDialect` class is modified to add the cast expression (e.g., `::uuid`) to the variable placeholder (e.g, `?`) if the column type matches one of the database types that requires casting, which in this case is either `uuid`, `json`, or `jsonb`.

In my test cases on PostgreSQL (12.6), inserts into `JSON`/`JSONB` columns worked fine without explicit casting, but I guess Confluent makes the assumption that if a column signals to be a `json` in `pg_catalog`, it would also support inserting values casted to `json`. Maybe it was required in older PostgreSQL versions to have that explicit cast.

**Comment by mfussenegger** [comment]
>  Maybe it was required in older PostgreSQL versions to have that explicit cast.

The [PR](https://github.com/confluentinc/kafka-connect-jdbc/pull/875) that introduced the commit you linked refers to the PostgreSQL 10 docs. Maybe PostgreSQL made it more resilient later.

Some options I can think of:

1. We add a `json` alias for object as initially proposed
2. We add a dedicated `json` type and allow implicit casts `text → json` and `json -> object`. Initially this type is not available for storage. Objects continue to be serialized as `json` via the postgres wire protocol.
3. We make a PR against kafka-connect-jdbc that removes the explicit cast requirement for json if the server version supports it. But we'd either have to make kafka-connect-jdbc CrateDB aware (We send a `crate_version` in the startup message, so that could be used to identify CrateDB. `asyncpg` is another client that does that). Otherwise we'd have to bump our emulated PostgreSQL version (currently at 10.5). This might have other implications - e.g. other clients could start to use different queries.
4. We make a PR against kafka-connect-jdbc to add an option that disables the explicit type casts, or makes it possible to override the list of types which require an explicit cast.

**Comment by proddata** [comment]
I'd vote for option 1 and later maybe add more functionality.

JSON per current spec (RFC8259) doesn't have to preserve an ordering, nor should it contain duplicate keys and the whitespace must be insignificant. I'd say it is fine if we deviate here from PostgreSQL. Also we could still implement sending a crate_version to acknowledge that, JSONs **might** have to be treated differently.

https://datatracker.ietf.org/doc/html/rfc8259

**Comment by mfussenegger** [comment]
>  I'd say it is fine if we deviate here from PostgreSQL

Do you have an argument for why it is fine to deviate in this case?
What if some other PostgreSQL application would depend on this behavior?

**Comment by dobe** [comment]
> In my test cases on PostgreSQL (12.6), inserts into `JSON`/`JSONB` columns worked fine without explicit casting, but I guess Confluent makes the assumption that if a column signals to be a `json` in `pg_catalog`, it would also support inserting values casted to `json`. Maybe it was required in older PostgreSQL versions to have that explicit cast.

[just tested on 9.6 ](http://sqlfiddle.com/#!17/7e29e/8/0) (the oldest supported version of pg) - the explicit cast is not required, so i do not see an obvious reason why this cast is needed. and if this is really not needed i think we could do a PR that just removes the explicit cast?

**Comment by proddata** [comment]
> Do you have an argument for why it is fine to deviate in this case?

As I said, is not part of RFC8259, which also PostgreSQL is referencing.

> In general, most applications should prefer to store JSON data as jsonb, unless there are quite specialized needs, such as legacy assumptions about ordering of object keys.


I see how CrateDB `OBJECT`s are closer to Postgres `JSONB` then `JSON`. 
What was the reason to use `JSON` over `JSONB` as return type?


> What if some other PostgreSQL application would depend on this behavior?

Right now it would not be compatible either. Imho this is legacy functionality, not following standards. Don't know if we closely have to replicate bad practice

> the explicit cast is not required

makes one question why that PR from 2020 exists in the first place?
Even the example given works:

http://sqlfiddle.com/#!15/6d6592/1

**Comment by amotl** [comment]
Hi again,

FWIW, I wanted to reference https://github.com/duckdb/duckdb/pull/1822 here, as DuckDB also improved support for using JSON-encapsulated data in SQL statements the other day.

With kind regards,
Andreas.

**Comment by hammerhead** [comment]
FYI, I was able to successfully test the new JSON data type with the Confluent JDBC Sink connector. An updated version of the Kafka Connect Howto is available as a draft for when 4.7 is released (crate/crate-howtos#290).
